onedal/datatypes/dlpack/data_conversion.cpp

```diff
@@ -0,0 +1,163 @@
+                                 ptr,
+                                 row_count,
+                                 col_count,
+                                 deleter,
+                                 std::vector<sycl::event>{},
+                                 layout);
+#else
+        throw std::invalid_argument(
+            "Input array located on a oneAPI device, but sklearnex installation does not have SYCL support.");
+#endif
+    }
+    else {
+        res = dal::homogen_table(ptr, row_count, col_count, deleter, layout);
+    }
+    return res;
+}
+
+dal::table convert_to_table(py::object obj, py::object q_obj) {
+    dal::table res;
+    bool versioned = false;
+    DLManagedTensor* dlm;
+    DLManagedTensorVersioned* dlmv;
+    DLTensor tensor;
+    dal::data_type dtype;
+
+    // extract __dlpack__ attribute from the input obj. This function should
+    // only be called if the attribute has been checked.
+    py::capsule caps = obj.attr("__dlpack__")();
+
+    // two different types of dlpack managed tensors are possible, with
+    // DLManagedTensor likely to be removed from future versions of dlpack.
+    // collect important aspects necessary for the macro offloading.
+    PyObject* capsule = caps.ptr();
+    if (PyCapsule_IsValid(capsule, "dltensor")) {
+        dlm = caps.get_pointer<DLManagedTensor>();
+        tensor = dlm->dl_tensor;
+    }
+    else if (PyCapsule_IsValid(capsule, "dltensor_versioned")) {
+        dlmv = caps.get_pointer<DLManagedTensorVersioned>();
+        if (dlmv->version.major > DLPACK_MAJOR_VERSION) {
+            throw std::runtime_error("dlpack tensor version newer than supported");
+        }
+        tensor = dlmv->dl_tensor;
+        versioned = true;
+    }
+    else {
+        throw std::runtime_error("unable to extract dltensor");
+    }
+
+    // Extract and convert a DLpack data type into a oneDAL dtype.
+    dtype = convert_dlpack_to_dal_type(tensor.dtype);
+
+    // if there is a queue, check that the data matches the necessary precision.
+#ifdef ONEDAL_DATA_PARALLEL
+    if (!q_obj.is(py::none()) && !q_obj.attr("sycl_device").attr("has_aspect_fp64").cast<bool>() &&
+        dtype == dal::data_type::float64) {
+        // If the queue exists, doesn't have the fp64 aspect, and the data is float64
+        // then cast it to float32 (using reduce_precision), error raised in reduce_precision
+        py::object copy = reduce_precision(obj);
+        res = convert_to_table(copy, q_obj);
+        return res;
+    }
+#endif // ONEDAL_DATA_PARALLEL
+
+    // unusual data format found, try to make contiguous, otherwise throw error
+    // error throw located in regenerate_layout
+    if (get_dlpack_layout(tensor) == dal::data_layout::unknown) {
+        // NOTE: this attempts to make a contiguous deep copy of the data
+        // if possible, this is expected to be a special case
+        py::object copy = regenerate_layout(obj);
+        res = convert_to_table(copy, q_obj);
```

Comment:
So this is some kind of a recursive call. Isn't it? Are there any limitations on the depth of recursion?

Reply:
As a quick fix I would recommend to add an assertion on object being convertible right copying it.